### PR TITLE
[JENKINS-22849] Improve HTTP response to include item queue location

### PR DESCRIPTION
**Pull request by ydubreuil**
[JENKINS-22849](https://issues.jenkins-ci.org/browse/JENKINS-22849)

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
:bee: , but there's one improvement for the original code.
<code>Jenkins.getInstance()</code> requires a null check before the schedule call. It's a mandatory behavior in these calls, because they may arrive when Jenkins starts a shutdown.

File: src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java [hunk 1]
```diff
@@ -83,8 +86,12 @@ public void doBuild(StaplerRequest req, StaplerResponse rsp, @QueryParameter Str
             LOGGER.fine("wrong kind");
             throw HttpResponses.error(HttpServletResponse.SC_BAD_REQUEST, "Use /buildByToken/buildWithParameters for this job since it takes parameters");
         }
-        Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), getBuildCause(req));
-        ok(rsp);
+        Queue.Item item = Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), getBuildCause(req));
+        if (item != null) {
+            rsp.sendRedirect(SC_CREATED, req.getContextPath() + '/' + item.getUrl());
+        } else {
+            rsp.sendRedirect(".");
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java]: sending a redirect on failure will be a success response - which seems incorrect here.

File: src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java [hunk 2]
```diff
@@ -83,8 +86,12 @@ public void doBuild(StaplerRequest req, StaplerResponse rsp, @QueryParameter Str
             LOGGER.fine("wrong kind");
             throw HttpResponses.error(HttpServletResponse.SC_BAD_REQUEST, "Use /buildByToken/buildWithParameters for this job since it takes parameters");
         }
-        Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), getBuildCause(req));
-        ok(rsp);
+        Queue.Item item = Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), getBuildCause(req));
+        if (item != null) {
+            rsp.sendRedirect(SC_CREATED, req.getContextPath() + '/' + item.getUrl());
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java]: Has this been tested - to redirect you would need a 301 or 302 response code - so `SC_CREATED` is incorrect or ignored.

File: src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java [hunk 3]
```diff
@@ -105,8 +112,12 @@ public void doBuildWithParameters(StaplerRequest req, StaplerResponse rsp, @Quer
         		values.add(value);
         	}
         }
-        Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), new ParametersAction(values), getBuildCause(req));
-        ok(rsp);
+        Queue.Item item = Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), new ParametersAction(values), getBuildCause(req));
+        if (item != null) {
```
Comment (jtnord) [inline src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java]: see comments above

**Comment by jtnord** [comment]
:bug: on the dubious redirects with status codes and redirect with success on a failure.

**Comment by ydubreuil** [comment]
@jtnord this is how it's currently implemented in Jenkins Core, see here: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/ParametersDefinitionProperty.java#L185-L192

I wanted to keep semantic of Jenkins core

**Comment by jtnord** [comment]
Hmm, so the javadoc for staplerResponse confused me and looking at the code it does the right thing.

So i am removing the ꞉bug꞉ - but I'm still dubious on the `sendRedirect(".")` despite it being an existing pattern - it seems like a bug to me, so can not in all consiousness vote  ꞉bee꞉

File: src/test/java/org/jenkinsci/plugins/build_token_root/BuildRootActionTest.java [hunk 4]
```diff
@@ -76,10 +76,10 @@
         wc.assertFails(p.getUrl() + "build?token=secret", HttpURLConnection.HTTP_FORBIDDEN);
         j.waitUntilNoActivity();
         assertEquals(0, p.getBuilds().size());
-        wc.goTo("buildByToken/build?job=" + p.getFullName() + "&token=secret&delay=0sec");
+        wc.goTo("buildByToken/build?job=" + p.getFullName() + "&token=secret&delay=0sec", null);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/build_token_root/BuildRootActionTest.java]: Please strengthen the test to prove that it is sending a 201 with a reasonable-looking URL.

**Comment by jglick** [comment]
> it seems like a bug to me, so can not in all consiousness vote ꞉bee꞉

@jtnord you can propose a (compatible!) improvement to core semantics, but this plugin is merely designed to mimic the core behavior with the only difference being permission handling.

> It's a mandatory behavior in these calls, because they may arrive when Jenkins starts a shutdown.

@oleg-nenashev if Jenkins were shutting down (or not yet started), it would not be servicing HTTP requests. (Yes there is a race condition there, but it affects practically every web method in Jenkins, and there is no useful recovery mechanism anyway; throwing a `NullPointerException` is as good as anything else.)

File: src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/build_token_root/BuildRootAction.java]: It is not really a “redirect”, but code 201 takes a `Location` header.

**Comment by jtnord** [comment]
> throwing a `NullPointerException` is as good as anything else

using a 500 is better :-)

**Comment by jglick** [comment]
> using a 500 is better

That is the code that actually gets sent in such cases IIRC.

**Comment by jtnord** [comment]
but with a stack trace rather than a human readable message.

**Comment by ydubreuil** [comment]
I can add this error handling, but I'm worried that it's not what people expect.  If people starts to use this plugin instead of core API after securing their Jenkins, they will expect to get the same behavior.

```
Jenkins j = Jenkins.getInstance();
if (j != null) {
    Queue.Item item = Jenkins.getInstance().getQueue().schedule(p, delay.getTime(), new ParametersAction(values), getBuildCause(req));
    if (item != null) {
        rsp.sendRedirect(SC_CREATED, req.getContextPath() + '/' + item.getUrl());
     } else {
        error(rsp, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Unable to schedule job");
     }
} else {
   error(rsp, HttpServletResponse.SC_SERVICE_UNAVAILABLE, "Jenkins is not ready");
}
```
@jglick  test strengthened

**Comment by jtnord** [comment]
>
> I can add this error handling, but I'm worried that it's not what people
> expect


I'm not voting a bug on the failure to create an item - I just could't vote
bee either - you don;t need all commentators to vote bee, just have no bugs.

I think catching the `Jenkins.getInstance() == null` and sending a 503 is a
good addition - the text could be clearer to say "Jenkins is starting up or
shutting down" (rather than just not ready)

**Comment by oleg-nenashev** [comment]
I also do not put a bug, but I thing it really makes sense to do. The time efforts would be less than the time, which has been already spent in this discussion ;)

**Comment by jglick** [comment]
`SC_INTERNAL_SERVER_ERROR` would certainly be wrong, as there are legitimate reasons for a failure to schedule, such as a `QueueDecisionHandler` rejection.

:bee: from me as it stands.

**Comment by jtnord** [comment]
:bee: I'm convinced

**Comment by ydubreuil** [comment]
@reviewbybees done

**Comment by ydubreuil** [comment]
I can squash my branch and rewrite git commit message to refer to issue JENKINS-22849 instead of JENKINS-29615 if needed

**Pull request merged**